serialization/serialization.go

```diff
@@ -0,0 +1,263 @@
+//
+// mysql::serialization is a serialization format introduced with tagged GTIDs
+//
+// https://dev.mysql.com/doc/dev/mysql-server/latest/PageLibsMysqlSerialization.html
+package serialization
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math/bits"
+	"strings"
+)
+
+// Message is a mysql::serialization message
+type Message struct {
+	Version uint8 // >= 0
+	Format  Format
+}
+
+func (m *Message) String() (text string) {
+	text += fmt.Sprintf("Message (version: %d)", m.Version)
+	for _, line := range strings.Split(m.Format.String(), "\n") {
+		text += "\n  " + line
+	}
+	return
+}
+
+// GetFieldByName returns a field if the name matches and an error if there is no match
+func (m *Message) GetFieldByName(name string) (Field, error) {
+	for _, f := range m.Format.Fields {
+		if f.Name == name {
+			return f, nil
+		}
+	}
+	return Field{}, fmt.Errorf("field not found: %s", name)
+}
+
+// Format is describing a `message_format`
+type Format struct {
+	Size                  uint64
+	LastNonIgnorableField int
+	Fields                []Field
+}
+
+func (f *Format) String() (text string) {
+	text += fmt.Sprintf("Format (Size: %d, LastNonIgnorableField: %d)\n",
+		f.Size, f.LastNonIgnorableField)
+	for _, f := range f.Fields {
+		text += fmt.Sprintf("Field %02d (Name: %s, Skipped: %t, Type: %T)\n",
+			f.ID, f.Name, f.Skipped, f.Type)
+		if f.Type != nil {
+			text += fmt.Sprintf("  Value: %s\n", f.Type.String())
+		}
+	}
+	return text
+}
+
+// Field represents a `message_field`
+type Field struct {
+	ID       int
+	Type     FieldType
+	Optional bool
+	Name     string
+	Skipped  bool
+}
+
+// FieldType represents a `type_field`
+type FieldType interface {
+	fmt.Stringer
+}
+
+// FieldIntFixed is for values with a fixed length.
+// This is also known as the 'fixlen_integer_format'.
+// The encoded value can vary be between 1 and 2 times
+// of that of the value before encoding.
+type FieldIntFixed struct {
+	Length int // Length of value before encoding, encoded value can be more
+	Value  []byte
+}
+
+func (f FieldIntFixed) String() string {
+	if f.Value == nil {
+		return ""
+	}
+	return fmt.Sprintf("0x%x", f.Value)
+}
+
+// FieldIntVar is using the signed integer variant of the 'varlen_integer_format'
+// and encodes a value as a byte sequence of 1-9 bytes depending on the value.
+type FieldIntVar struct {
+	Value int64
+}
+
+func (f FieldIntVar) String() string {
+	return fmt.Sprintf("%d", f.Value)
+}
+
+// FieldUintVar is using the usigned integer variant of the 'varlen_integer_format'
+// and encodes a value as a byte sequence of 1-9 bytes depending on the value.
+type FieldUintVar struct {
+	Value uint64
+}
+
+func (f FieldUintVar) String() string {
+	return fmt.Sprintf("%d", f.Value)
+}
+
+// FieldString is a 'string_format' field
+type FieldString struct {
+	Value string
+}
+
+func (f FieldString) String() string {
+	return f.Value
+}
+
+func Unmarshal(data []byte, v interface{}) error {
```

Comment:
it's better to split this function into two dedicated functions for Message and Format? I don't understand why they are merged. seems not shared many logic

Reply:
If you compare it to `json.Unmarshal()` that also doesn't have different functions depending on the type (array, string, object, etc).
I think I like to keep it like this.